libs/dyn/convert/normalize.go

```diff
@@ -74,30 +74,32 @@ func (n normalizeOptions) normalizeStruct(typ reflect.Type, src dyn.Value, seen
 
 	switch src.Kind() {
 	case dyn.KindMap:
-		out := make(map[string]dyn.Value)
+		out := dyn.NewMapping()
 		info := getStructInfo(typ)
-		for k, v := range src.MustMap() {
-			index, ok := info.Fields[k]
+		for _, pair := range src.MustMapping().Pairs() {
+			pk := pair.Key
+			pv := pair.Value
+			index, ok := info.Fields[pk.MustString()]
 			if !ok {
 				diags = diags.Append(diag.Diagnostic{
 					Severity: diag.Warning,
-					Summary:  fmt.Sprintf("unknown field: %s", k),
-					Location: src.Location(),
+					Summary:  fmt.Sprintf("unknown field: %s", pk.MustString()),
+					Location: pk.Location(),
```

Comment:
This previously pointed to the location of the parent map value. Now it points to the unknown key.